apollo-query-graph/src/extract_subgraphs_from_supergraph.rs

```diff
@@ -0,0 +1,1703 @@
+use apollo_at_link::federation_schema::{
+    FederationSchema, FederationSchemaRef, OptionLinksMetadata,
+};
+use apollo_at_link::federation_spec_definition::{FederationSpecDefinition, FEDERATION_VERSIONS};
+use apollo_at_link::join_spec_definition::{
+    FieldDirectiveArguments, JoinSpecDefinition, TypeDirectiveArguments, JOIN_VERSIONS,
+};
+use apollo_at_link::link_spec_definition::LinkSpecDefinition;
+use apollo_at_link::location::{
+    InputObjectTypeDefinitionLocation, InterfaceTypeDefinitionLocation,
+    ObjectTypeDefinitionLocation, TypeDefinitionLocation, UnionTypeDefinitionLocation,
+};
+use apollo_at_link::spec::{Identity, Version};
+use apollo_at_link::spec_definition::SpecDefinition;
+use apollo_compiler::ast::FieldDefinition;
+use apollo_compiler::schema::{
+    Component, ComponentOrigin, ComponentStr, DirectiveDefinition, DirectiveLocation, Directives,
+    EnumType, EnumValueDefinition, ExtendedType, ExtensionId, InputObjectType,
+    InputValueDefinition, InterfaceType, Name, NamedType, ObjectType, ScalarType, Type, UnionType,
+};
+use apollo_compiler::{Node, NodeStr, Schema};
+use apollo_federation_error::error::{ErrorCode, FederationError};
+use indexmap::{IndexMap, IndexSet};
+use lazy_static::lazy_static;
+use std::collections::BTreeMap;
+use std::ops::Deref;
+
+// Assumes the given schema has been validated.
+#[allow(dead_code)]
+fn extract_subgraphs_from_supergraph(
+    supergraph_schema: &Schema,
+    validate_extracted_subgraphs: Option<bool>,
+) -> Result<FederationSubgraphs, FederationError> {
+    let validate_extracted_subgraphs = validate_extracted_subgraphs.unwrap_or(true);
+    let (supergraph_schema, link_spec_definition, join_spec_definition) =
+        validate_supergraph(supergraph_schema)?;
```

Comment:
somewhat curious -> why would we allow passing invalid schema in the first place? maybe this method should be on `Supergraph` to ensure schema is validated?

Reply:
FWIW the function is named `validate_supergraph()` in the JS code, but it does very little actual validation on the supergraph schema (it's closer to an `extract_specs()` function than a `validate_supergraph()` function). We don't really have code in the `federation` repo that fully validates a supergraph schema, but rather code that tries to use the supergraph schema to do things (and if there are invalidities discovered along the way, it panics or errors).

We could certainly move this function to `Supergraph` if you'd like, and move the function to a method there as well (and have this `extract_subgraphs_from_supergraph()` function take the `Supergraph` struct instead of `Schema`). I can also rename the function to `extract_specs()` or something. Would that be fine?